File: src/EvernightAI/core/schema/memory.py
from enum import StrEnum
from datetime import datetime, timezone
from typing import Any

from pydantic import Field, field_validator, model_validator

from EvernightAI.core.schema.base import EvernightAISchema


class MemoryKind(StrEnum):
    """记忆类型"""

    FACT = "fact"
    PREFERENCE = "preference"
    SUMMARY = "summary"
    DEFINITION = "definition"
    INSTRUCTION = "instruction"
    EPISODIC = "episodic"


class MemoryScope(StrEnum):
    """记忆作用域"""

    GLOBAL = "global"
    USER = "user"
    SESSION = "session"
    CONTEXT = "context"


class MemorySort(StrEnum):
    """记忆排序方式"""

    DEFAULT = "default"
    PRIORITY = "priority"
    RELEVANCE = "relevance"
    CONFIDENCE = "confidence"
    UPDATED_AT = "updated_at"
    CREATED_AT = "created_at"
    MEMORY_ID = "memory_id"


class MemoryWriteOperation(StrEnum):
    """记忆写入操作"""

    CREATE = "create"
    REPLACE = "replace"
    MERGE = "merge"


class MemoryScopeSelector(EvernightAISchema):
    """记忆作用域选择器"""

    scope: MemoryScope
    scope_id: str | None = None

    @field_validator("scope_id")
    @classmethod
    def _normalize_scope_id(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    @model_validator(mode="after")
    def _validate_scope_identity(self) -> "MemoryScopeSelector":
        if self.scope is MemoryScope.GLOBAL and self.scope_id is not None:
            raise ValueError("global memory selector must not have a scope_id")
        if self.scope is not MemoryScope.GLOBAL and self.scope_id is None:
            raise ValueError("non-global memory selector must have a scope_id")
        return self


class MemoryItem(EvernightAISchema):
    """记忆项"""

    memory_id: str
    owner_id: str | None = None
    content: str
    kind: MemoryKind = MemoryKind.FACT
    scope: MemoryScope = MemoryScope.GLOBAL
    scope_id: str | None = None
    tags: list[str] = Field(default_factory=list)
    priority: int = 0
    relevance: float = Field(default=0.0, ge=0.0, le=1.0)
    confidence: float = Field(default=1.0, ge=0.0, le=1.0)
    is_enabled: bool = True
    expires_at: datetime | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("content")
    @classmethod
    def _content_must_not_be_empty(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("memory content must not be empty")
        return normalized

    @field_validator("scope_id", "owner_id")
    @classmethod
    def _normalize_optional_identity(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    @field_validator("tags")
    @classmethod
    def _normalize_tags(cls, value: list[str]) -> list[str]:
        normalized: list[str] = []
        seen: set[str] = set()
        for tag in value:
            next_tag = tag.strip().casefold()
            if not next_tag or next_tag in seen:
                continue
            seen.add(next_tag)
            normalized.append(next_tag)
        return normalized

    @field_validator("expires_at", "created_at", "updated_at")
    @classmethod
    def _normalize_datetime(cls, value: datetime | None) -> datetime | None:
        if value is None:
            return None
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("memory datetime values must be timezone-aware")
        return value.astimezone(timezone.utc)

    @model_validator(mode="after")
    def _validate_scope_identity(self) -> "MemoryItem":
        if self.scope is MemoryScope.GLOBAL and self.scope_id is not None:
            raise ValueError("global memory must not have a scope_id")
        if self.scope is not MemoryScope.GLOBAL and self.scope_id is None:
            raise ValueError("non-global memory must have a scope_id")
        if self.updated_at < self.created_at:
            raise ValueError("memory updated_at must not be before created_at")
        return self


class MemoryQuery(EvernightAISchema):
    """记忆查询"""

    text: str | None = None
    scope: MemoryScope | None = None
    scope_id: str | None = None
    scopes: list[MemoryScopeSelector] = Field(default_factory=list)
    kinds: list[MemoryKind] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    minimum_relevance: float | None = Field(default=None, ge=0.0, le=1.0)
    minimum_confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    include_disabled: bool = False
    include_expired: bool = False
    deduplicate: bool = False
    sort: MemorySort = MemorySort.DEFAULT
    limit: int | None = Field(default=None, ge=1)
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("text")
    @classmethod
    def _normalize_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = " ".join(value.split())
        return normalized or None

    @field_validator("scope_id")
    @classmethod
    def _normalize_scope_id(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    @field_validator("tags")
    @classmethod
    def _normalize_tags(cls, value: list[str]) -> list[str]:
        normalized: list[str] = []
        seen: set[str] = set()
        for tag in value:
            next_tag = tag.strip().casefold()
            if not next_tag or next_tag in seen:
                continue
            seen.add(next_tag)
            normalized.append(next_tag)
        return normalized


class MemorySelection(EvernightAISchema):
    """记忆选择结果"""

    memories: list[MemoryItem] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)

File: frontend/src/api/memory.ts
import { requestJson } from './client'

export type MemoryKind =
  | 'fact'
  | 'preference'
  | 'summary'
  | 'definition'
  | 'instruction'
  | 'episodic'

export type MemoryScope = 'global' | 'user' | 'session' | 'context'
export type MemorySort =
  | 'default'
  | 'priority'
  | 'relevance'
  | 'confidence'
  | 'updated_at'
  | 'created_at'
  | 'memory_id'

export type MemoryItem = {
  memory_id: string
  owner_id?: string | null
  content: string
  kind?: MemoryKind
  scope?: MemoryScope
  scope_id?: string | null
  tags?: string[]
  priority?: number
  relevance?: number
  confidence?: number
  is_enabled?: boolean
  expires_at?: string | null
  created_at?: string
  updated_at?: string
  metadata?: Record<string, unknown>
}

export type MemoryQuery = {
  text?: string | null
  scope?: MemoryScope | null
  scope_id?: string | null
  kinds?: MemoryKind[]
  tags?: string[]
  minimum_relevance?: number | null
  minimum_confidence?: number | null
  include_disabled?: boolean
  include_expired?: boolean
  deduplicate?: boolean
  sort?: MemorySort
  limit?: number | null
  metadata?: Record<string, unknown>
}

export type MemorySelection = {
  memories: MemoryItem[]
  metadata?: Record<string, unknown>
}

export function createMemory(memory: MemoryItem): Promise<MemoryItem> {
  return requestJson<MemoryItem>('/memories', {
    method: 'POST',
    body: memory,
  })
}

export type MemoryListParams = {
  text?: string
  scope?: MemoryScope | ''
  scope_id?: string
  kind?: MemoryKind | ''
  tag?: string
  include_disabled?: boolean
  include_expired?: boolean
  deduplicate?: boolean
  sort?: MemorySort
}

export function listMemories(params: MemoryListParams = {}): Promise<MemoryItem[]> {
  const searchParams = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value === undefined || value === null || value === '') {
      return
    }
    searchParams.append(key, String(value))
  })
  const query = searchParams.toString()
  return requestJson<MemoryItem[]>(query ? `/memories?${query}` : '/memories')
}

export function getMemory(memoryId: string): Promise<MemoryItem> {
  return requestJson<MemoryItem>(`/memories/${encodeURIComponent(memoryId)}`)
}

export function selectMemories(query: MemoryQuery | null = null): Promise<MemorySelection> {
  return requestJson<MemorySelection>('/memories/select', {
    method: 'POST',
    body: query,
  })
}

export function replaceMemory(memoryId: string, memory: MemoryItem): Promise<MemoryItem> {
  return requestJson<MemoryItem>(`/memories/${encodeURIComponent(memoryId)}`, {
    method: 'PUT',
    body: memory,
  })
}

export function enableMemory(memoryId: string): Promise<MemoryItem> {
  return requestJson<MemoryItem>(`/memories/${encodeURIComponent(memoryId)}/enable`, {
    method: 'POST',
  })
}

export function disableMemory(memoryId: string): Promise<MemoryItem> {
  return requestJson<MemoryItem>(`/memories/${encodeURIComponent(memoryId)}/disable`, {
    method: 'POST',
  })
}

export function deleteMemory(memoryId: string): Promise<void> {
  return requestJson<void>(`/memories/${encodeURIComponent(memoryId)}/delete`, {
    method: 'POST',
  })
}
